acnportal/acnsim/models/battery.py

```diff
@@ -131,10 +132,105 @@ class Linear2StageBattery(Battery):
+                'Voltage must be greater than 0. Got {0}'.format(voltage))
+        if period <= 0:
+            raise ValueError(
+                'period must be greater than 0. Got {0}'.format(period))
+        if pilot == 0:
+            self._current_charging_power = 0
+            return 0
+        # All calculations are done in terms of battery SoC, so we
+        # convert pilot signal and max power into pilot and max rate of
+        # change of SoC.
+        pilot_dsoc = pilot * voltage / 1000 / self._capacity / (60 / period)
+        max_dsoc = self._max_power / self._capacity / (60 / period)
+
+        if pilot_dsoc > max_dsoc:
+            pilot_dsoc = max_dsoc
+
+        # The pilot SoC rate of change has a new transition SoC at˚
+        # which decreasing of max charging rate occurs.
+        pilot_transition_soc = (
+            self._transition_soc
+            + (pilot_dsoc - max_dsoc) / max_dsoc * (self._transition_soc - 1)
+        )
+
+        if pilot < 0:
+            warnings.warn(f"Negative pilot signal input. Battery models"
+                          f"may not be accurate for pilot {pilot} A.")
+
+        # The charging equation depends on whether the current SoC of
+        # the battery is above or below the new transition SoC.
+        if self._soc < pilot_transition_soc:
+            # In the pre-rampdown region, the charging equation changes
+            # depending on whether charging the battery over this
+            # time period causes the battery to transition between
+            # charging regions.
+            if 1 <= (pilot_transition_soc - self._soc) / pilot_dsoc:
+                curr_soc = pilot_dsoc + self._soc
+            else:
+                curr_soc = (
+                    1
+                    + np.exp(
+                        (pilot_dsoc + self._soc - pilot_transition_soc)
+                        / (pilot_transition_soc - 1)
+                    )
+                    * (pilot_transition_soc - 1)
+                )
+        else:
+            curr_soc = (
+                1
+                + np.exp(pilot_dsoc / (pilot_transition_soc - 1))
+                * (self._soc - 1)
+            )
+
+        # Add subtractive noise to the final SoC, scaling the noise
+        # such that _noise_level is the standard deviation of the noise
+        # in the battery charging power.
+        if self._noise_level > 0:
+            raw_noise = np.random.normal(0, self._noise_level)
+            scaled_noise = raw_noise * (period / 60) / self._capacity
+            curr_soc -= abs(scaled_noise)
+
+        dsoc = curr_soc - self._soc
+        self._current_charge = curr_soc * self._capacity
+
+        # For charging power and charging rate (current), we use the
+        # the average over this time period.
+        self._current_charging_power = dsoc * self._capacity / (period / 60)
+        return self._current_charging_power * 1000 / voltage
+
+    def charge_old(self, pilot, voltage, period):
```

Comment:
I had originally included the old charge function just in case; I didn't have a specific use case in mind. Do we still want to include the old charge function for comparison/legacy purposes? If so, we should include functionality to access it outside the battery.py file. Something like a `use_old_charge_func` flag and a wrapper `charge()` method that uses either the new `charge` (or `_charge`) function or the `_old_charge` function depending on the flag. I can do this if it's something we want. Otherwise I recommend removing `charge_old`.

Personally, I think we should keep charge_old for now, so that users of the previous version of the L2S battery still have a way to replicate their results. What do you think? @zach401

Reply:
I agree. We should probably keep the charge_old method and make it an option. We should come up with a more descriptive name / flag name though. Another option would be to use inheritance so that one class uses the old charge method and one uses the new. What do you think?

Reply:
How about `charge_constant_in_timestep`? Kind of long, but says what the old charging method does: within a single timestep, charges at a constant rate. If we want something shorter, maybe `charge_discrete` or `charge_stepwise`? As for the flag name, something like `use_constant_timestep_charging` maybe?

I don't think we should use inheritance to expose the new charge method. As far as I can tell, there's no reason for a new user of acnportal to use the old charging method, as the new method is more accurate in all cases. We're only including the old charge method for legacy reproducibility reasons, not for any physical reason.

Reply:
I think I like charge_stepwise. Perhaps the flag in the charge method could be something like charge_calculation='stepwise'. And for the new method it would be change_calculation='continuous'. This would allow us to have other calculation methods if it ever makes sense. Its similar to the syntax for picking the type of plot  or interpolation method in pandas.

Reply:
Or maybe calculation_method as the argument?

Reply:
I like `charge_stepwise` for the method and `charge_calculation` for the flag. I think `charge_calculation` is clearer than `calculation_method` since `charge_calculation` is clear about what exactly we're calculating: charge as opposed to noise, etc.